### ISSUE TITLE
Race condition in C# initialization code

### ISSUE THREAD

**Issue by Compynerd255** [ValveSoftware/openvr#449]
If multiple threads are using the OpenVR interface in .NET, there is a possibility of a race condition when any of the VR components (such as the compositor) are being initialized. Specifically, it is possible for two copies of any of these components to be created and for certain threads to get stale copies, because two threads can both end up in the "create component if null" conditional if they both express interest in that interface around the same time:

    public CVRCompositor VRCompositor()
    {
        CheckClear();
        if (m_pVRCompositor == null)
        {
            var eError = EVRInitError.None;
            var pInterface = OpenVRInterop.GetGenericInterface(FnTable_Prefix+IVRCompositor_Version, ref eError);
            if (pInterface != IntPtr.Zero && eError == EVRInitError.None)
                m_pVRCompositor = new CVRCompositor(pInterface);
            }
        return m_pVRCompositor;
    }

**Comment by destrofer** [comment]
Since you are using threads you should never forget about thread synchronization (https://msdn.microsoft.com/en-us/library/mt679037.aspx). Simplest way would be to use `lock`:

```C#
private static bool CompositorLock = false;

public CVRCompositor VRCompositor()
{
	CheckClear();
	lock(CompositorLock)
	{
		if (m_pVRCompositor == null)
		{
			var eError = EVRInitError.None;
			var pInterface = OpenVRInterop.GetGenericInterface(FnTable_Prefix+IVRCompositor_Version, ref eError);
			if (pInterface != IntPtr.Zero && eError == EVRInitError.None)
				m_pVRCompositor = new CVRCompositor(pInterface);
		}
	}
	return m_pVRCompositor;
}
```

The **static** field `CompositorLock` is required since `lock` cannot be applied on null objects.

You also should not forget to make sure that your CheckClear() method is multithread-friendly too.

Making OpenVR multithread-friendly might not be the best thing to do since most applications use it only in one single thread and needless thread synchronization would create a computing overhead. So it really should be up to your application to make sure threads are friendly with each other and share OpenVR toys without tearing them apart :)

**Comment by ahtur** [comment]
> Yes, I agree that multithreading is probably an edge case (though I honestly think that fully parallelized engines will be the norm in a few years), but it would be nice to have at least some documentation that openvr_api.cs is not thread safe, and perhaps provide a version that is.
> 
> In fact, since it only appears to be OpenVR object initialization that is not thread safe, you could probably use a function pointer to overwrite the getter methods so that the versions with the locks (the one that creates the objects) are only called once.

Agreed. An official statement about multithreading in openvr would be much appreciated.
